driver-core/src/main/com/mongodb/internal/connection/SaslAuthenticator.java

```diff
@@ -131,6 +112,48 @@ private void throwIfSaslClientIsNull(final SaslClient saslClient) {
         }
     }
 
+    private BsonDocument getNextSaslResponse(final SaslClient saslClient, final InternalConnection connection) {
+        BsonDocument response = getSpeculativeAuthenticateResponse();
+        if (response != null) {
+            return response;
+        }
+
+        try {
+            byte[] serverResponse = saslClient.hasInitialResponse() ? saslClient.evaluateChallenge(new byte[0]) : null;
+            return sendSaslStart(serverResponse, connection);
+        } catch (Exception e) {
+            throw wrapException(e);
+        }
+    }
+
+    private void getNextSaslResponseAsync(final SaslClient saslClient, final InternalConnection connection,
+                                          final SingleResultCallback<Void> callback) {
+        BsonDocument response = getSpeculativeAuthenticateResponse();
+        try {
+            if (response == null) {
+                byte[] serverResponse = (saslClient.hasInitialResponse() ? saslClient.evaluateChallenge(new byte[0]) : null);
+                sendSaslStartAsync(serverResponse, connection, new SingleResultCallback<BsonDocument>() {
```

Comment:
I would recommend wrapping the callback `new SingleResultCallback<BsonDocument>() {``in a `ErrorHandlingResultCallback` so that any errors resulting when processing the of this callback would be handled.